File: experiments/experiment2/sources/factorize/factorize.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

// factorize は試し割り法で素因数分解を行う
func factorize(n int64) {
	fmt.Printf("%d: ", n)

	if n < 2 {
		fmt.Printf("%d\n", n)
		return
	}

	temp := n

	// 2で割れるだけ割る
	for temp%2 == 0 {
		fmt.Print("2 ")
		temp /= 2
	}

	// 3以上の奇数で試し割り（√temp まで）
	for i := int64(3); i*i <= temp; i += 2 {
		for temp%i == 0 {
			fmt.Printf("%d ", i)
			temp /= i
		}
	}

	// 最後に残った数が 2 以上なら素数
	if temp > 1 {
		fmt.Printf("%d", temp)
	}
	fmt.Println()
}

func main() {
	// 引数チェック
	if len(os.Args) < 2 {
		fmt.Printf("Usage: %s <num1> <num2> ... <numN>\n", os.Args[0])
		return
	}

	for _, arg := range os.Args[1:] {
		// 文字列を int64 に変換
		n, err := strconv.ParseInt(arg, 10, 64)
		if err != nil || n <= 0 {
			fmt.Printf("%s: Positive integer required.\n", arg)
			continue
		}
		factorize(n)
	}
}

File: experiments/experiment2/sources/factorize/factorize.c
#include <stdio.h>
#include <stdlib.h>

/**
 * 試し割り法による素因数分解
 * @param n 分解する数値
 */
void factorize(long long n) {
    printf("%lld: ", n);
    
    if (n < 2) {
        printf("%lld\n", n);
        return;
    }

    // 2で割れるだけ割る
    while (n % 2 == 0) {
        printf("2 ");
        n /= 2;
    }

    // 3以上の奇数で試し割り（√n まで）
    for (long long i = 3; i * i <= n; i += 2) {
        while (n % i == 0) {
            printf("%lld ", i);
            n /= i;
        }
    }

    // 最後に残った数が 2 以上なら、それは素数
    if (n > 1) {
        printf("%lld", n);
    }
    printf("\n");
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <num1> <num2> ... <numN>\n", argv[0]);
        return 1;
    }

    for (int i = 1; i < argc; i++) {
        // 文字列を数値に変換 (long long型)
        long long n = atoll(argv[i]);
        if (n <= 0) {
            printf("%s: Positive integer required.\n", argv[i]);
            continue;
        }
        factorize(n);
    }

    return 0;
}

